Sources/XiEditor/LineCache.swift

```diff
@@ -170,7 +174,20 @@ fileprivate class LineCacheState<T>: UnfairLock {
                     }
                     let startIx = oldIx - nInvalidBefore
                     if op_type == "copy" {
-                        newLines.append(contentsOf: lines[startIx ..< startIx + nCopy])
+                        var lineNumber = op["ln"] as! UInt
+                        let toCopy = lines[startIx ..< startIx + nCopy]
+                        // the line number in the update is the logical line number of the first VISUAL
+                        // line to copy. If this line is not itself logical, increment lineNumber for the next line.
+                        if toCopy.first??.number == nil {
```

Comment:
Am I missing something about Swift syntax or should this be `first?.number`?

Reply:
Hmm, or is this actually `toCopy.first ?? (self).number`? In that case, please reformat.

Reply:
no, this is a double optional; the items in this list are themselves option types, and the `first` accessor returns an option (the list may be empty). I did a double take as well.

Reply:
Care to add a comment :D I'm sure I'll be confused again next week if I get back to this code.